src/maxdiffusion/models/train.py

```diff
@@ -353,12 +366,13 @@ def compute_loss(unet_params):
     mllog_utils.train_init_stop(config)
     mllog_utils.train_run_start(config)
     mllog_utils.train_step_start(config, start_step)
-    for step in np.arange(start_step, config.max_train_steps):
+    for step in np.arange(start_step, config.max_train_steps+1):
```

Comment:
config.max_train_steps+1 will make train loop train extra batch. the total training step would be max_train_step - start_step + 2

Reply:
np.arange is exclusive of config.max_train_steps, and it seems state.step starts at 1? I found through running the pipeline multiple times that we were excluding the last step, thus I added it.

Reply:
Actually I verified that state.step starts at 0, but even with this logic and running training for 21 steps,   completed step is correct, so perhaps there is a bug somewhere else.

Ex:

```bash
python src/maxdiffusion/models/train.py src/maxdiffusion/configs/base15.yml cache_dir=gs://jfacevedo-maxdiffusion/cache_dir per_device_batch_size=8 run_name="sd15_test_main_with_ckpts_2" base_output_directory=gs://jfacevedo-maxdiffusion/training_results/ attention=flash checkpoint_every=256 upload_ckpts_to_gcs=True max_train_steps=21
```

logs:

```bash
completed step: 0, seconds: 13.993, TFLOP/s/device: 1.378, loss: 0.061
To see full metrics 'tensorboard --logdir=sd-model-finetuned/sd15_test_main_with_ckpts_2/tensorboard/'
completed step: 1, seconds: 0.190, TFLOP/s/device: 101.697, loss: 0.053
completed step: 2, seconds: 0.185, TFLOP/s/device: 104.269, loss: 0.043
completed step: 3, seconds: 0.159, TFLOP/s/device: 120.980, loss: 0.055
completed step: 4, seconds: 0.261, TFLOP/s/device: 73.988, loss: 0.052
completed step: 5, seconds: 0.260, TFLOP/s/device: 74.092, loss: 0.064
completed step: 6, seconds: 14.093, TFLOP/s/device: 1.368, loss: 0.040
completed step: 7, seconds: 0.029, TFLOP/s/device: 669.515, loss: 0.054
completed step: 8, seconds: 20.464, TFLOP/s/device: 0.942, loss: 0.055
completed step: 9, seconds: 0.033, TFLOP/s/device: 577.246, loss: 0.047
completed step: 10, seconds: 0.232, TFLOP/s/device: 83.278, loss: 0.070
completed step: 11, seconds: 0.261, TFLOP/s/device: 73.772, loss: 0.057
completed step: 12, seconds: 0.261, TFLOP/s/device: 73.992, loss: 0.041
completed step: 13, seconds: 0.260, TFLOP/s/device: 74.135, loss: 0.072
completed step: 14, seconds: 0.261, TFLOP/s/device: 73.770, loss: 0.079
completed step: 15, seconds: 0.261, TFLOP/s/device: 73.978, loss: 0.058
completed step: 16, seconds: 20.706, TFLOP/s/device: 0.931, loss: 0.057
completed step: 17, seconds: 0.029, TFLOP/s/device: 660.797, loss: 0.052
completed step: 18, seconds: 0.236, TFLOP/s/device: 81.676, loss: 0.053
completed step: 19, seconds: 0.259, TFLOP/s/device: 74.365, loss: 0.048
completed step: 20, seconds: 0.262, TFLOP/s/device: 73.662, loss: 0.045
Moving file metrics_step_000000_to_step_000020.txt to GCS...
File metrics_step_000000_to_step_000020.txt moved successfully!
/home/jfacevedo/miniconda3/envs/maxdiffusion/lib/python3.10/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
loading configuration file config.json from cache at /home/jfacevedo/.cache/huggingface/hub/models--CompVis--stable-diffusion-safety-checker/snapshots/cb41f3a270d63d454d385fc2e4f571c487c253c5/config.json
`text_config_dict` is provided which will be used to initialize `CLIPTextConfig`. The value `text_config["id2label"]` will be overriden.
`text_config_dict` is provided which will be used to initialize `CLIPTextConfig`. The value `text_config["bos_token_id"]` will be overriden.
`text_config_dict` is provided which will be used to initialize `CLIPTextConfig`. The value `text_config["eos_token_id"]` will be overriden.
`text_config` is `None`. Initializing the `CLIPTextConfig` with default values.
`vision_config` is `None`. initializing the `CLIPVisionConfig` with default values.
Model config CLIPConfig {
  "_name_or_path": "clip-vit-large-patch14/",
  "architectures": [
    "SafetyChecker"
  ],
  "initializer_factor": 1.0,
  "logit_scale_init_value": 2.6592,
  "model_type": "clip",
  "projection_dim": 768,
  "text_config": {
    "dropout": 0.0,
    "hidden_size": 768,
    "intermediate_size": 3072,
    "model_type": "clip_text_model",
    "num_attention_heads": 12
  },
  "torch_dtype": "float32",
  "transformers_version": "4.40.2",
  "vision_config": {
    "dropout": 0.0,
    "hidden_size": 1024,
    "intermediate_size": 4096,
    "model_type": "clip_vision_model",
    "num_attention_heads": 16,
    "num_hidden_layers": 24,
    "patch_size": 14
  }
}

loading weights file pytorch_model.bin from cache at /home/jfacevedo/.cache/huggingface/hub/models--CompVis--stable-diffusion-safety-checker/snapshots/cb41f3a270d63d454d385fc2e4f571c487c253c5/pytorch_model.bin
Loading PyTorch weights from /home/jfacevedo/.cache/huggingface/hub/models--CompVis--stable-diffusion-safety-checker/snapshots/cb41f3a270d63d454d385fc2e4f571c487c253c5/pytorch_model.bin
PyTorch checkpoint contains 303,981,845 parameters.
...
```

Reply:
This seems a bug with our logger. printing step vs our logger:

```bash
step:  0
step:  1
completed step: 0, seconds: 36.641, TFLOP/s/device: 0.526, loss: 0.061
To see full metrics 'tensorboard --logdir=sd-model-finetuned/sd15_test_main_with_ckpts_2/tensorboard/'
```

I'll work on a fix.

Reply:
I made a fix to this, write_metrics needs to be called at the end before saving the weights.